### ISSUE TITLE
Move regex/wildcard path handlers to base class

### ISSUE THREAD

**Issue by shravan2x** [unosquare/embedio#98]
I am using EmbedIO 1.9.1 and have code similar to this:

```c#
_httpServer = WebServer.Create($"http://localhost:8080");
_httpServer.RegisterModule(new TestModule());
```

```c#
class TestModule : WebModuleBase
{
    public TestModule()
    {
        AddHandler("/settings", HttpVerbs.Get, OnFetchSettings);
        AddHandler("/users/*", HttpVerbs.Post, OnCreateUser);
    }
    
    // OnFetchSettings and other handlers
}
```

When run, the first handler is works correctly, but not the second. That is, the correct handler is called for `/settings`, but `/users/37` results in a 404.

Note that I am not using the WebApiModule. I assume that the default matcher includes support for wildcards since it is accepted as an argument to the `WebServer` constructor itself, rather than `WebApiModule`.

**Comment by geoperez** [comment]
The WebModuleBase path matcher is not using wildcards. This functionality is only present in WebApiModule. You may need to implement it by yourself.

**Comment by shravan2x** [comment]
Oh I see. Is this the same case for the Regex matching handler as well?

**Comment by geoperez** [comment]
Yes, you need to implement it.

**Comment by shravan2x** [comment]
Okay, I guess I was just confused. Is there any particular reason this is taken as a parameter to `WebServer` rather than `WebApiModule` itself?

**Comment by geoperez** [comment]
Not really, I guess to allow more modules use them. If you think the current path handler from WebAPI can be moved to the base class let me know.

**Comment by shravan2x** [comment]
Yeah, I think that moving the path handlers to the base class would be a wise choice.

- It will reduce confusion like in my case, where I spent almost an hour debugging the problem myself before posting the issue here and learning that it isn't used at all.
- It will increase compatibility with the `*` handler. The current behavior here is confusing since the wildcard works when used in `*`, but something like `/path/*` does not work.
- More modules can use this built-in functionality without having to reinvent the wheel each time.
- And most importantly, users can easily to use it without having to set up an entire MVC pattern. I usually prefer EmbedIO over ASP.NET in cases where I need a lightweight HTTP server. But setting up an entire MVC pattern here leads to the same level of work as ASP, which is counterintuitive. And I usually end up adding an `*` handler and writing my own path handler in every application.

I think these are the biggest reasons to move core path handling functionality to the base class. Further, we also have a `RoutingStrategy.Plain` which only does 1:1 string equality matches (maybe even a `RoutingStrategy.None`, which does not match any routes and allows the user to add modules to do so).

**Comment by mariodivece** [comment]
@shravan2x I think these are very good recommendations!

**Comment by geoperez** [comment]
@shravan2x what do you think? https://github.com/unosquare/embedio/pull/99

**Comment by shravan2x** [comment]
@geoperez I think the new changes look really good. I plan to build and test it on my projects tomorrow.

Is there any way we can warn users when they select `Regex` handling in the base class without a module that handles it? If not, maybe we could mention it in the docs.

**Comment by geoperez** [comment]
I can't see where to warn about the routing strategy, because they can add a new module in any moment.

**Comment by shravan2x** [comment]
One idea might be when they call `RunAsync()` to start the server, since it is a reasonable assumption that modules are usually added before that.

--------------------------------------

I feel having the Regex option when it doesn't actually do anything without an _optional_ module is confusing behavior. We should ideally maintain a generic base with modules that add specializations. Can we do one of:

- Implement Regex handling in the base class (imo this would be more useful and it won't result in a performance hit unless users select the option anyway).
- Remove the Regex option and add it as native functionality to the `WebApiModule` instead?

**Comment by geoperez** [comment]
Regex handling is the best idea. Let me think about it.

**Comment by geoperez** [comment]
@shravan2x The Regex handler is ready to be used in base class. Take a look in the branch.

**Comment by shravan2x** [comment]
I spent some time testing the new features. Everything looks good from here!

**Comment by geoperez** [comment]
Excellent, branch merged, releasing version 1.10 and closing issue.
